net, codec: Extract a Datagram framed implementation for UDP and UDS

Kobzol opened the pull request:
This PR attempts to extract a common datagram frame implementation and use it for both UDP and UDS (as suggested by: https://github.com/tokio-rs/tokio/issues/1458).

Right now it's only implemented for UDP.

Questions:
1) Where should the implementation of `DatagramSocket` for UDP and UDS live? Right now it's in the `tokio-net` crate to use the crate private implementations of `UdpSocket`.
2) UDSFramed was not working before this change (it was commented out and using old futures). Should I implement equivalents of `poll_send_to_priv` and `poll_recv_from_priv` found in `UdpSocket` or should this be solved with some public API (this would also allow to implement it directly in `tokio-codec`, as stated in the previous question).

This PR will probably conflict with:
https://github.com/tokio-rs/tokio/pull/1445

carllerche commented on tokio-codec/src/datagram.rs:
```diff
@@ -0,0 +1,179 @@
+    flushed: bool,
+}
+
+pub trait DatagramSocket<A> {
```
I believe `A` should be an associated type. We could call this `Address`?

I also don't think `tokio-net` should have a dependency on `tokio-codec`, so maybe lets move this trait into `tokio-io` and name it `AsyncDatagram` (matching `AsyncRead` / `AsyncWrite`).

Kobzol replied:
You're right, that should absolutely be an associated type :) I moved `AsyncDatagram` to `tokio-io`.

carllerche commented on tokio-net/src/udp/socket.rs:
```diff
@@ -355,6 +356,16 @@ impl fmt::Debug for UdpSocket {
     }
 }
 
+impl DatagramSocket<SocketAddr> for UdpSocket {
```
The impl is fine here.

Kobzol commented:
I included the bound `S: AsyncDatagram` directly in `DatagramFramed`'s impl, so that it is not even possible to construct it with a type that doesn't implement it. Otherwise it could be misleading that you can construct `DatagramFramed` with any type, but it wouldn't have `Stream` and `Sink` implemented in such a case. It would be nice to also do it with the codecs, but I would need to say something like `C: Encoder OR Decoder`, which is not possible to express in Rust right now I think.

I'll work on porting UDS now.

andrewtj commented:
It doesn't look like UDP sockets bound to `ANY` fit in here. For them you need to capture the destination at receive time and set that as the source when you reply.

Kobzol commented:
I think that this doesn't fit into the datagram `Stream`/`Sink` scenario. You would need some trait that asynchronously receives a datagram and then returns you the datagram along with an object that can be used for sending data. This could be probably added for both UDP and UDS, but I don't think that it necessarily fits here, this is a trait for simple (unconnected) datagram sending and receiving.

Kobzol commented:
I implemented `AsyncDatagram` for `UnixDatagram` and added a test for framed UDS. Let me know if you want me to change anything, otherwise only documentation is missing I think, but I wanted to keep that until the end, once we're sure that we're happy with the implementation.

andrewtj commented:
@Kobzol if it doesn't fit that's okay, but I don't quite follow what you've written. Specifically I'm not sure what you're referring to by object. The only difference is returning and taking two `Address` at send and receive time. Also this isn't to do with connected sockets, multi-homed servers bound to `ANY` aren't that unusual, and they will want to reply to clients from the correct address.

Kobzol commented:
Ah, I probably misunderstood what you want. Could you please explain to me a bit how this works? I didn't find any methods on `UDPSocket` (in `tokio`/`mio`/`std`) that would allow sending a datagram while taking two addresses (source + dest). Do you have to (re)bind before sending each datagram?

Couldn't it be done like this (half-pseudocode)?
```rust
let socket = UDPSocket...;
loop {
     let (dgram, addr) = socket.next().await;
     let response_socket = UDPSocket::bind(addr); // bind to some address
     response_socket.send(...).await;
}
```

andrewtj commented:
There's no built-in support. To get and set the source/received address you have to bind with some extra options and then get or set the address via control messages ([this article](https://blog.powerdns.com/2012/10/08/on-binding-datagram-udp-sockets-to-the-any-addresses/comment-page-1/) goes into more detail).

I like to think this is fundamental UDP functionality, but realistically no else has raised this (as far as I can tell), and I've dropped the ball in not doing so myself when I rolled my own implementation.

Kobzol commented:
Oh, I see. In that case this it would probably be better to first add support for this in `std` and/or `mio`, because that's where the actual implementation lives, `tokio` is just an async wrapper on top of that. If this existed in `mio`, there could be another trait that would work with address pairs.

andrewtj commented:
> there could be another trait that would work with address pairs.

Two traits is what I was hoping to avoid by raising this.

Kobzol commented:
This feature is not yet implemented anywhere and so we're not sure of the semantics, therefore I don't think that we should block this trait for it. Also not every datagram has to support this (is this valid for Unix sockets), so I would still make a different trait for it. @carllerche what do you think about this?

LucioFranco commented:
I think it may still make sense to provide a trait like this but this PR has gotten stale and most likely we should probably add this to `tokio-util`. I am going to close this and if you feel that is incorrect feel free to reopen it.
